framework/src/userobject/GhostingUserObject.C

```diff
@@ -0,0 +1,107 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "GhostingUserObject.h"
+#include "NonlinearSystem.h"
+#include "MooseMesh.h"
+
+registerMooseObject("MooseApp", GhostingUserObject);
+
+namespace
+{
+constexpr unsigned int GEOMETRIC_MAP_IDX = 0;
+constexpr unsigned int ALGEBRAIC_MAP_IDX = 1;
+}
+
+template <>
+InputParameters
+validParams<GhostingUserObject>()
+{
+  InputParameters params = validParams<GeneralUserObject>();
+
+  params.addParam<std::vector<processor_id_type>>("pids", "The PID(s) to see the ghosting for");
+
+  // No need to set or override this flag, this object is triggered by meshChanged events
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_INITIAL};
+
+  params.addClassDescription("Creates ghosting maps that can be queried by external objects.");
+  return params;
+}
+
+GhostingUserObject::GhostingUserObject(const InputParameters & parameters)
+  : GeneralUserObject(parameters), _mesh(_subproblem.mesh()), _nl(_fe_problem.getNonlinearSystem())
+{
+  _maps.resize(2); // Geometric and Algebraic maps
+
+  if (parameters.isParamValid("pids"))
+    _pids = getParam<std::vector<processor_id_type>>("pids");
+  else
+  {
+    // If a PIDs vector is not passed, generate patterns for all processors
+    _pids.resize(_app.n_processors());
+    std::iota(_pids.begin(), _pids.end(), 0);
+  }
+}
+
+void
+GhostingUserObject::initialSetup()
+{
+  meshChanged();
+}
+
+void
+GhostingUserObject::meshChanged()
+{
+  // Clear the maps before rebuilding
+  for (auto & map_ref : _maps)
+    for (auto pid_map_pair : map_ref)
+      pid_map_pair.second.clear();
+
+  for (auto pid : _pids)
+  {
+    auto begin_elem = _mesh.getMesh().active_pid_elements_begin(pid);
+    const auto end_elem = _mesh.getMesh().active_pid_elements_end(pid);
+
+    // Loop over all the geometric ghosting functors and build up a map of elements that are ghosted
+    // for each PID
+    libMesh::GhostingFunctor::map_type geometric_elems;
+    for (auto & gf : as_range(_mesh.getMesh().ghosting_functors_begin(),
+                              _mesh.getMesh().ghosting_functors_end()))
+      (*gf)(begin_elem, end_elem, pid, geometric_elems);
+    _maps[GEOMETRIC_MAP_IDX].emplace(pid, geometric_elems);
+
+    // Loop over all the algebraic ghosting functors and build up a map of elements that are ghosted
+    // for each PID
+    libMesh::GhostingFunctor::map_type algebraic_elems;
+    for (auto & gf : as_range(_nl.dofMap().algebraic_ghosting_functors_begin(),
```

Comment:
I do feel like we kind of have a void throughout our whole RelationshipManager system with respect to `coupling_functors`. In `DofMap::add_neighbors_to_send_list` we loop over both the algebraic and coupling functors to build the ghosting for the `DofMap`. Perhaps we should loop over the coupling functors here too just to make sure we don't miss any ghosting? In MOOSE we seem to be treating algebraic and coupling ghosting as the same thing, which conceptually I am ok with...treating the system vectors and matrices as a unit. So I think including ghosting of coupling functors in our "algebraic" ghosting would be fine. What do you think?

Reply:
Sure - I haven't given enough thought to the cases where the distinction will matter. Since they both get attached to the DOF map, we may not need the extra confusion at the MOOSE level. I'll add the loop here though for now. Later we can always split it out if people want to see that field separately. I don't believe we've added a single Coupling functor to any of the existing MOOSE systems yet though so that can wait.